nav2_graceful_motion_controller/include/nav2_graceful_motion_controller/graceful_motion_controller.hpp

```diff
@@ -0,0 +1,179 @@
+   * @param velocity  Current robot velocity
+   * @param goal_checker Ptr to the goal checker for this task in case useful in computing commands
+   * @return          Best command
+   */
+  geometry_msgs::msg::TwistStamped computeVelocityCommands(
+    const geometry_msgs::msg::PoseStamped & pose,
+    const geometry_msgs::msg::Twist & velocity,
+    nav2_core::GoalChecker * goal_checker) override;
+
+  /**
+   * @brief nav2_core setPlan - Sets the global plan.
+   * @param path The global plan
+   */
+  void setPlan(const nav_msgs::msg::Path & path) override;
+
+  /**
+   * @brief Limits the maximum linear speed of the robot.
+   * @param speed_limit expressed in absolute value (in m/s)
+   * or in percentage from maximum robot speed
+   * @param percentage setting speed limit in percentage if true
+   * or in absolute values in false case
+   */
+  void setSpeedLimit(const double & speed_limit, const bool & percentage) override;
+
+protected:
+  /**
+   * @brief Get motion target point.
+   * @param motion_target_dist Optimal motion target distance
+   * @param path Current global path
+   * @return Motion target point
+   */
+  geometry_msgs::msg::PoseStamped getMotionTarget(
+    const double & motion_target_dist,
+    const nav_msgs::msg::Path & path);
+
+  /**
+   * @brief Simulate trajectory calculating in every step the new velocity command based on
+   * a new curvature value and checking for collisions.
+   *
+   * @param robot_pose Robot pose
+   * @param motion_target Motion target point
+   * @param trajectory Simulated trajectory
+   * @param backward Flag to indicate if the robot is moving backward
+   * @return true if the trajectory is collision free, false otherwise
+   */
+  bool simulateTrajectory(
+    const geometry_msgs::msg::PoseStamped & robot_pose,
+    const geometry_msgs::msg::PoseStamped & motion_target, nav_msgs::msg::Path & trajectory,
+    const bool & backward);
+
+  /**
+   * @brief Rotate the robot to face the motion target with maximum angular velocity.
+   *
+   * @param angle_to_target Angle to the motion target
+   * @return geometry_msgs::msg::Twist Velocity command
+   */
+  geometry_msgs::msg::Twist rotateToTarget(
+    const double & angle_to_target);
+
+  /**
+   * @brief Checks if the robot is in collision
+   * @param x The x coordinate of the robot in global frame
+   * @param y The y coordinate of the robot in global frame
+   * @param theta The orientation of the robot in global frame
+   * @return Whether in collision
+   */
+  bool inCollision(const double & x, const double & y, const double & theta);
+
+  rclcpp_lifecycle::LifecycleNode::WeakPtr parent_node_;
+  std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
+  std::string plugin_name_;
+  std::shared_ptr<nav2_costmap_2d::Costmap2DROS> costmap_ros_;
+  std::unique_ptr<nav2_costmap_2d::FootprintCollisionChecker<nav2_costmap_2d::Costmap2D *>>
+  collision_checker_;
+  geometry_msgs::msg::TransformStamped costmap_transform_;
```

Comment:
This could be passed to the simulate function without the need for storing as a class member murking up the logic flow, since its only used twice